### ISSUE TITLE
question on building stLinkHack

### ISSUE THREAD

**Issue by franksterb92** [GMMan/st-link-hack#1]
im trying to build the stlinkhack and I'm getting 11 errors that deviceManager doesn't exist and windowsUsbDeviceFactory couldn't be found and filterDeviceDefinition doesn't contain a definition among others

**Comment by GMMan** [comment]
Make sure you're using the specific version of Device.NET as referenced in the submodules.

**Comment by franksterb92** [comment]
ok ill have to check it out later but was def wanting to copy some stlink bootloaders tho just in case. idk why st dont just share it i mean i get it tho  but china offers a way cheaper programmer for them anyway what version is it 4

**Comment by franksterb92** [comment]
i got it build now just to figure out the other steps lol imma buy a few cheap nucleo boards so i can spare a few to brick lmao

**Comment by Bob-the-Kuhn** [comment]
I was able to get it to build.  The steps I took were (this is on a Windows PC):

1. Download & unzip the code for the **stlink-v2 branch**
2. Navigate to the file **StLinkHack.sln**
3. Right click **StLinkHack.sln** and select **open with** > **Micosoft Studio 2022**
4. This brings up a fresh Micosoft Studio project.  Before building you need to select the CPU.  On the second line from the top are two boxes. One says **Debug** and the other says **Any CPU**.  Click the **Any CPU** box and select **New**.
5. That pops up the **Configuration Manager**.  Click in the **Active solution platform** box and select **New**.
6. In the **New Solution Platform** popup click the **Type or select new platform** box and select the CPU of your PC.  In the **Copy settings from** box click **empty**.   Leave the **Create new project platforms** box checked. and click **OK**.
7. The **Configuration Manager** should now show the new CPU for all three projects.  Close the **Configuration Manager** popup.
8. You can now build the solution.
9. Finding the EXE file is "fun".  On my machine it was in the directory **st-link-hack\host_app\StLinkHack\bin\x64\Debug\netcoreapp3.1**

---

I made the mistake of creating a new Visual Studio project and then copying in the StLinkHack code.  Chased my tail for better that a week before starting over from scratch.

---

I have not been able to download the StLink firmware.  It hangs at the **Sending payload...** step.

I've tried 0x200007d4, 0x20001c9c, 0x20001c9e, 0x20001cA0 and 0x20001c9a.

It would be really nice if someone posted a downloaded STLink image.

**Comment by GMMan** [comment]
Did you build the payload?

**Comment by Bob-the-Kuhn** [comment]
Yes, I built the payload.

I did get the **"Waiting for start sequence..."** prompt at the following offsets but no activity on the TX pin.
0x200007e0
0x200007d6
0x200007ce
0x200007a6

**Bad news** - I tanked two STLinks while trying to download from them.  I was trying different offsets when they happened.  I think one of them is 0x200007a4.  Unfortunately I closed the shell before I recorded the address.

My testing consisted of:

1. Copying payload.bin into the same directory as the StLinkHack executable
2. Opening a powershell window
3. Navigating to the StLinkHack executable directory
4. Executing variations of **./StLinkHack.exe  payload.bin COM25 0x200007d4 output.bin**
5. I tried 0x200007a4 through 0x20000800.

The STLinks I tanked were M1137 and MB1136 flavors.

**Comment by GMMan** [comment]
It seems the boards you're using contain ST-Link/V2-1, not ST-Link/V2. The branch `stlink-v2` is specifically for standalone ST-Link/V2 and clones, and it has a different bootloader with different addresses compared to V2-1.

**Comment by Bob-the-Kuhn** [comment]
Rats

**Comment by GMMan** [comment]
If you have another spare ST-Link/V2-1, try the master branch and dump the bootloader. You should be able to recover the other ones with the same bootloader, and you can use STLinkUpgrade to do a complete installation. You'll have to manually adjust the volume label if you want them to go back to saying NUCLEO though.

**Comment by Bob-the-Kuhn** [comment]
The master branch is known to be compatible with the ST-Link/V2-1?

**Comment by GMMan** [comment]
Yes. If you read the upgrade guide, you'll see the branches it suggests for different versions.

**Comment by Bob-the-Kuhn** [comment]
It's amazing what happens when you follow the instructions!!!

I was able to download the image from one of my STLink_V2 clones.

---

I think where I screwed up is I used the payload and offset for the STLink_V2  on my STLink_V2-1 units.

I have a new unit coming in so maybe I can get it right tomorrow.

**Comment by Bob-the-Kuhn** [comment]
I was able to download a V2-1 from a nucleo-F302R8. 

[STLinkV2_1_V2.J33.M25.bin.zip](https://github.com/GMMan/st-link-hack/files/12320514/STLinkV2_1_V2.J33.M25.bin.zip)

I directly programmed a nucleo-F746ZG with it and all I got was a blinking red LED.

I then did the clone changes and got the same results.

Suggestions?

FYI - the magic value is different on this image than shown in your write up.  On this image it is 0x47a53c15.  It also shows up at offset 0xA328.  I'm guessing that the difference x47 vs. xA7 is actually the hardware version.

**Comment by GMMan** [comment]
Can you try writing just the bootloader and see if it will show up?

**Comment by Bob-the-Kuhn** [comment]
No luck with the bootloader only images.  

I was able to download the modified file to one of my STLink-V2 clones and upgrade it.

Strange - works on a clone but not a real one.

**Comment by GMMan** [comment]
> FYI - the magic value is different on this image than shown in your write up. On this image it is 0x47a53c15. It also shows up at offset 0xA328. I'm guessing that the difference x47 vs. xA7 is actually the hardware version.

I think that was just a typo. I dug out my original images and it's 0x47a53c15.

> I directly programmed a nucleo-F746ZG with it and all I got was a blinking red LED.

That means it isn't connected to USB. Check your cable?

**Comment by Bob-the-Kuhn** [comment]
Finally got my nucleo-F746ZG fully functional.

The image was one I downloaded from a nucleo-F413ZH with the two volume names changed.

To program the nucleo-F746ZG I did the following:

1. Moved zero ohm resistor SB108 to SB109 (swdio)
2. Moved zero ohm resistor SB103 to SB104 (swclk)
3. Plugged in a STLink to CN6 (SWD connector)
4. Provide power either via CN6 (SWD connector) or via a power only USB cable.   It won't connect with a fully functional USB cable atteched the the target's USB port.
5. Connected to the target. 
6. Changed ROP to disabled (which erases the chip)
7. Programmed the image

To make the board functional I reversed the zero ohm resistor changes.

I then used STLinkUpgrade.jar to upgrade to the latest firmware.  After that I could use the STLink utility
